## Introdução à Linguagem

### Comentários em Python

In [1]:
# Este é um comentário
# Toda função tem parênteses e print() é uma função (função escreva).
print("Teste")

Teste


In [6]:
"""
Este é um comentário
Mas desta vez
É um comentário longo
"""
print("Teste")

Teste


### O ponto e vírgula (;)
O ponto e vírgula no Python não é obrigatório a menos que queira executar duas funções que estão na mesma linha

In [8]:
print("Teste")
print("Teste")

Teste
Teste


In [7]:
print("Teste"); print("Teste")

Teste
Teste


### Identação
A identação é obrigatória quando houver hierarquia

In [5]:
if True:
    print("Isto é verdadeiro")
else:
    print("Isto é falso")
print("Isto está fora e vai ser impresso")

Isto é verdadeiro
Isto está fora e vai ser impresso


### Variáveis
Espaços na memória em que serão armazenados dados

Toda variável é um OBJETO para o Python

In [1]:
nome = "Icaro" # String = str
idade = 38 # inteiro = int
peso = 83.5 # Python utiliza padrão americano por isso usa-se ponto (flutuante = float) ao invés de vírgula
altura = 1.78 # float = float
situacao = True # valor lógico, booleano = bool
print(nome)
type(peso)

Icaro


float

In [7]:
nada = None # variável sem nenhum valor associado
print(nada)
print(type(nada))

None
<class 'NoneType'>


In [1]:
# declarando duas variáveis ou mais na mesma linha
x, y = 20, 30
print(x)
print(y)

20
30


## Manipulação de Dados

### Saída de Dados

In [2]:
print("Olá Mundo!") # print é uma função (leva parênteses)

Olá Mundo!


In [4]:
print("Olá \nMundo!") # o \n quebra linha

Olá 
Mundo!


In [67]:
print(7+4) # soma inteiros
print("7"+"4") # concatena strings

11
74


In [68]:
print("Olá" + 5) # Erro

TypeError: can only concatenate str (not "int") to str

In [71]:
print("Olá", 5) # a vírgula concatena tipos diferentes

Olá 5


In [82]:
# print de variáveis
print(nome)
print(peso)
print(idade)

Icaro
83.5
38


In [3]:
# concatenação
print("Meu nome é", nome, "meu peso é", peso, "e minha idade é", idade, "anos")

Meu nome é Icaro meu peso é 83.5 e minha idade é 38 anos


In [4]:
# saída formatada
print("Meu nome é {}, meu peso é {} e minha idade é {} anos".format(nome, peso, idade)) # forma antiga
print(f"Meu nome é {nome}, meu peso é {peso} e minha idade é {idade} anos") # a partir do python 3.7

Meu nome é Icaro, meu peso é 83.5 e minha idade é 38 anos
Meu nome é Icaro, meu peso é 83.5 e minha idade é 38 anos


In [3]:
# acrescentando número de caracteres
print("Meu nome é {:20}!".format(nome)) # acrescenta espaços para completar 20 caracteres
print(f"Meu nome é {nome:20}!") # python 3.7

# alinhamento
print(f"Meu nome é {nome:>20}!") #à direita
print(f"Meu nome é {nome:^20}!") #centralizado
print(f"Meu nome é {nome:.<20}!") #à esquerda e o "." substitui os espaços 
print(f"Meu nome é {nome:=^20}!")

Meu nome é Icaro               !
Meu nome é Icaro               !
Meu nome é                Icaro!
Meu nome é        Icaro        !
Meu nome é Icaro...............!
Meu nome é =======Icaro========!


In [3]:
# separador
print("Dia", "Mês", "Ano", sep = "/") # o espaço (padrão) é substituído pela "/"
print('14', '02', '1984', sep = "/")
print('1984', '02', '14', sep = '-')
print('B', 'n', 'n', '.', sep = 'a')

Dia/Mês/Ano
14/02/1984
1984-02-14
Banana.


In [4]:
# finalização de linha
print("Vamos estudar", end = ' ') # substitui o "\n" (padrão) que é a quebra de linha
print("Python!")
print("Preço: ", end = 'R$')
print(50.0)

Vamos estudar Python!
Preço: R$50.0


In [2]:
print('quebra'); print('linha')

print('mesma', end= ' ')
print('linha')

quebra
linha
mesma linha


### Entrada de Dados

In [7]:
nome = input()

João


In [8]:
print(nome)

João


O padrão de Entrada de Dados é do tipo primitivo String, sendo necessário converter (cast) em alguns casos.

In [12]:
nome = input("Digite seu nome: ")
idade = int(input("Digite sua idade: ")) #cast
peso = float(input("Digite seu peso: ")) #cast
print(f"A variável nome é do tipo {type(nome)}")
print(f"A variável idade é do tipo {type(idade)}")
print(f"A variável peso é do tipo {type(peso)}")
print(f"Meu nome é {nome}, meu peso é {peso} e minha idade é {idade} anos")

Digite seu nome: Icaro
Digite sua idade: 32
Digite seu peso: 62.5
A variável nome é do tipo <class 'str'>
A variável idade é do tipo <class 'int'>
A variável peso é do tipo <class 'float'>
Meu nome é Icaro, meu peso é 62.5 e minha idade é 32 anos


In [14]:
teste = 'Icaro10'
print(f"A variável teste é do tipo {type(teste)}")

# spoiler: em Python toda variável é um objeto e objetos podem possuir funções/métodos internos
# para objetos do tipo string, há métodos built-in de teste de tipo como:
print(teste.isnumeric()) # é numérico?
print(teste.isalpha()) # é alfabético?
print(teste.isalnum()) # é alfanumérico?
print(teste.isspace()) # só tem espaço?
print(teste.isupper()) # está em maiúsculo?
print(teste.islower()) # está em minúsculo?
print(teste.istitle()) # está capitalizado?
# spoiler: um objeto tem atributos e métodos, mas como tem parênteses ao final, neste caso são métodos.

A variável teste é do tipo <class 'str'>
False
False
True
False
False
False
True


<mark> Fazer os exercícios do 1 ao 4. </mark>

### Operadores Aritméticos
Adição: +

Subtração: -

Divisão: /

Multiplicação: *

Potência: **

Divisão inteira: //

Resto da divisão (módulo): %

In [5]:
a = 4
b = 4
print(a + b)
print(5 - 2)
print(a * 5)
print(a**2)
print(10 / 6)
print(10 // 6)
print(10 % 2) # útil para separar pares (0) e ímpares (1)

8
3
20
16
1.6666666666666667
1
0


In [15]:
'''
Ordem de Precedência:
1. ()
2. **
3. *, /, //, % (o que aparecer primeiro)
4. +, -
'''
print(4+6*4/(4-2)**2)
print(4+6*4/pow((4-2),2))

10.0
10.0


In [5]:
# formatação pré-definida
a = 2
b = 3
print(a/b)
print('%f'%(a/b)) #padrão de 6 casas decimais
print('%.2f'%(a/b)) #2 casas decimais (modo mais antigo)

0.6666666666666666
0.666667
0.67


In [22]:
div = a/b
print('{:.2f}'.format(div)) # ainda antigo
print(f'{div:.2f}') # Python 3.7
print(f'{div:=>20.2f}')

0.67
0.67
================0.67


In [9]:
# operando com strings
print('Oi'*5)
print('='*20)

OiOiOiOiOi


<mark> Fazer os exercícios do 5 ao 15. </mark>

## Strings

### Introdução a strings (cadeias de caracteres)

In [42]:
nome = "Icaro Martins"
nome2 = 'Icaro Martins' # tanto faz aspas duplas ou simples (mais usual)
nome3 = '''
    Esta é uma string longa
    contendo informações sobre
    Python
'''
print(nome3) # o espaço com TAB (margem) também é impresso


    Esta é uma string longa
    contendo informações sobre
    Python



### Propriedades das strings

In [11]:
# spoiler: strings são listas de caracteres

lista = 'lista de letras'
tamanho = len(lista) # função para ver comprimento de listas
print(tamanho)

15


In [3]:
contagem = lista.count('a') # função Built-in: método da classe string que conta a quantidade de letras 'a's
print(contagem)

2


In [4]:
procurar = lista.find('de') # método que diz a posição do index em que começa
procurar2 = lista.find('python') # quando não encontra, aparece -1
print(procurar)
print(procurar2)

6
-1


In [5]:
# uso do in em listas (parecido com conjunto = set)

existe = 'lista' in lista
nexiste = 'python' in lista
print(existe)
print(nexiste)

True
False


In [6]:
# transformação

substituir = lista.replace('letras', 'python') # é possível alterar lista através de método
print(substituir)

maiusculo = lista.upper() # lower para minúsculo
print(maiusculo)

localiza = lista.upper().find('LETRAS') # transforma em maiúsculo e procura a string em maiúsculo
print(localiza)

capitalizar = lista.capitalize() # só a primeira em maiúsculo
print(capitalizar)

titulo = lista.title() # primeira letra de cada palavra em maiúsculo
print(titulo)

lista de python
LISTA DE LETRAS
9
Lista de letras
Lista De Letras


In [12]:
print(type(lista))
lista = lista.split(sep=' ') # divide string onde tiver espaços criando uma lista que contém strings
print(lista)
print(type(lista))

<class 'str'>
['lista', 'de', 'letras']
<class 'list'>


In [10]:
frase = '-'.join(lista) # junção da lista em uma frase
print(frase)

lista-de-letras


In [13]:
data = '14/2/1984'
lista2 = data.split('/') # não precisa escrever "sep" para indicar o parâmetro, porém é melhor explícito do que implícito.
print(lista2)

['14', '2', '1984']


In [14]:
frase = '   Aprenda    Python  '
cortar = frase.strip() # retira espaços do início e fim
print(len(frase))
print(cortar)
print(len(cortar))
cortar2 = frase.rstrip() # corta à direita (right)
cortar3 = frase.lstrip() # corta à esquerda (left)
print(cortar2)
print(len(cortar2))
print(cortar3)
print(len(cortar3))

22
Aprenda    Python
17
   Aprenda    Python
20
Aprenda    Python  
19


### Fatiando strings

In [15]:
#       012345
nome = 'python'
nova_string = nome[1] # indexação de string
nova_string2 = nome[-6] # conta "de trás pra frente" (na verdade -6 é menor que -5, etc.)
nova_string3 = nome[1:4] # igual a [1:4), ou seja, não imprime o 4º.
nova_string4 = nome[:5] # começa do 0, ou seja, é igual a [0:5)
nova_string5 = nome[3:] # começa do 3 e vai até o final.
nova_string6 = nome[0:4:2] # [0,4), de 2 em 2.
print(nova_string)
print(nova_string2)
print(nova_string3)
print(nova_string4)
print(nova_string5)
print(nova_string6)
print(lista[2]) # index 2 da lista
print(lista[2][3]) # index 3 da string no index 2 da lista.

y
p
yth
pytho
hon
pt
letras
r


In [20]:
# string, apesar de ser uma lista de caracteres, é imutável

nome[0] = 'P' # erro, pois só pode alterar com o método nome.replace()

TypeError: 'str' object does not support item assignment

In [19]:
# contagem com fatiamento
       # 012345
lista = 'lista de letras'
contagemf = lista.count('a',0,5)
print(contagemf)

1


<mark> Fazer os exercícios do 16 ao 21. </mark>

## Estruturas de Condição

### Operadores de Relação e Condição
Operadores de igualdade: != e ==

Operadores relacionais: >, <, >= e <=

In [17]:
numero1 = 100 == 100 # lê-se apenas um "=" como "recebe"
numero2 = 100 != 100
numero3 = 100 >= 100
numero4 = 100 > 100

print(numero1)
print(numero2)
print(numero3)
print(numero4)
type(numero1)

True
False
True
False


bool

In [2]:
idade = int(input('Digite sua idade: '))
resp = idade >= 18 # comparação que gera resposta booleana

print(resp)
type(resp)

Digite sua idade: 17
False


bool

### Comando if else

In [3]:
# condição simples (if apenas)

if resp == True: # o "T" de "True" deve ser digitado em maiúsculo!
    print('Você já é maior de idade') # não escreverá nada se for menor de idade

In [4]:
# condição composta (if e else)

if resp == True:
    print('Você já pode dirigir.')
else:
    print('Você não pode dirigir!')

Você não pode dirigir!


In [13]:
senha_cadastrada = 123
senha_digitada = 123
pode_acessar = senha_digitada == senha_cadastrada

if pode_acessar: # fazer pergunta ao ler: pode acessar? 
    print('Acesso permitido')
else:
    print('Acesso negado')

Acesso permitido


In [20]:
numero = 11

if numero == 10: # True
    print(f"Este número é igual a {numero}")
else: # False
    print("Este número não é igual a 10")

Este número não é igual a 10


In [21]:
# condição simplificada (na mesma linha)

print(f"Este número é igual a {numero}" if numero == 10 else "Este número não é igual a 10")

Este número não é igual a 10


In [22]:
idade = 40

if idade >= 20:
    if idade <= 60:
        print('A idade está entre 20 e 60 anos')
    else:
        print('A idade é maior que 60 anos')
else:
    print('A idade é menor que 20 anos')

A idade está entre 20 e 60 anos


In [23]:
# and

if idade >=20 and idade <=60:
    print('A idade está entre 20 e 60 anos')
else:
    print('A idade não está entre 20 e 60 anos') # é menor que 20 ou maior que 60

A idade está entre 20 e 60 anos


In [24]:
# or
if idade < 20 or idade > 60:
    print('A idade não está entre 20 e 60 anos')
else:
    print('A idade está entre 20 e 60 anos') # é maior ou igual a 20 e menor ou igual a 60

A idade está entre 20 e 60 anos


<mark> Fazer os exercícios do 22 ao 29. </mark>

### Comando elif

In [26]:
numero = 13

# condições aninhadas
if numero == 10: # True
    print(f"Este número é igual a {numero}")
elif numero == 11: # True
    print(f"Este número é igual a {numero}")
elif numero == 12: # True
    print(f"Este número é igual a {numero}")
elif numero == 13: # True
    print(f"Este número é igual a {numero}")
else: # False
    print("Este número não é igual a 10, 11, 12 ou 13")

Este número é igual a 13


In [13]:
nome = str(input('Qual é o seu nome? '))

if nome == 'Icaro':
    print('Que nome bonito!')
elif nome == 'Pedro' or nome == 'Paulo':
    print('Seu nome é popular no Brasil!')
elif nome in 'Ana Cláudia Jéssica Juliana': # dentro (in) da lista
    print('Belo nome feminino!')
#else:
    #print('Seu nome é bem normal!')
print(f'Tenha um bom dia, {nome}!')

Qual é o seu nome? Jéssica
Belo nome feminino!
Tenha um bom dia, Jéssica!


In [6]:
# 2 ifs != if e elif

x = 10

if x == 10:
    x = 11
if x == 11:
    x = 12
print(x) # vai printar 12

x = 10

if x == 10:
    x = 11
elif x == 11:
    x = 12
print(x) # vai printar 11, pois a primeira condição já foi satisfeita

12
11


<mark> Fazer os exercícios do 30 ao 39. </mark>

## Extra: Adicionando CORES no Python

### Padrão ANSI - padrão de normalização internacional

Código de escape sequence: \033[  m

In [110]:
# códigos para estilo: 0 (nenhum), 1(negrito), 4(sublinhado) e 7(negativo, cores invertidas)
# códigos para cor do texto: 30(branco), 31(vermelho), 32(verde), 33(amarelo), 34(azul), 35(roxo), 36(ciano), 37(cinza)
# códigos para cor do background: 40(branco), 41(vermelho), 42(verde), 43(amarelo), 44(azul), 45(roxo), 46(ciano), 47(cinza)
# estilo;cor do texto;cor do background
# exemplo: \033[0;33;44m
print('\033[7;32;43mBRASIL!')


BRASIL!


## Estruturas de Controle de Fluxo

### Estrutura de Repetição for
Estrutura de repetição com variável de controle

In [8]:
# laço com variável de controle

for c in range(1, 6): # c é a variável de controle e range é o intervalo [1,6)
    print('Oi ->', end= ' ')
print('FIM')

print()

for c in range(0, 6): # intervalo [0, 6)
    print(f'{c} ->', end= ' ')
print('FIM')

Oi -> Oi -> Oi -> Oi -> Oi -> FIM

0 -> 1 -> 2 -> 3 -> 4 -> 5 -> FIM


In [9]:
for c in range(6, 0):
    print(c) # Não printa, pois o padrão é somar 1
print('FIM')

print()

for c in range(6, 0, -1): # conta subtraindo 1
    print(f'{c} ->', end= ' ')
print('FIM')

print()

for c in range(0, 7, 2): # intervalo [0,7), somando 2
    print(f'{c} ->', end= ' ')
print('FIM')

FIM

6 -> 5 -> 4 -> 3 -> 2 -> 1 -> FIM

0 -> 2 -> 4 -> 6 -> FIM


In [10]:
n = int(input('Digite um número: '))

for c in range(0,n+1): # intervalo [0,n+1) ou seja, intervalo [0,n]
    print(f'{c} ->', end= ' ')
print('FIM')

print()

i = int(input('Início: '))
f = int(input('Fim: '))
p = int(input('Passo: '))

for c in range(i, f+1, p): # intervalo [i,f], somando p
    print(f'{c} ->', end= ' ')
print('FIM')

Digite um número: 3
0 -> 1 -> 2 -> 3 -> FIM

Início: 2
Fim: 19
Passo: 3
2 -> 5 -> 8 -> 11 -> 14 -> 17 -> FIM


In [11]:
soma = 0

for c in range(0, 4): # intervalo [0,3]
    num = int(input('Digite um valor: '))
    soma += num # soma = soma + num
print(f'O somatório de todos os valores foi {s}')

Digite um valor: 3
Digite um valor: 2
Digite um valor: 4
Digite um valor: 6
O somatório de todos os valores foi 15


In [17]:
for c in range(6): # range(0,6) = [0, 5]
    print(f'{c} ->', end= ' ')
print('FIM')

print()
       # 0123456789..
frase = 'curso de Python'
print(f'-=-=-= A frase possui {len(frase)} letras -=-=-=')
print('contagem: ', end= '')
for c in range(len(frase)):
    print(f'{c} ', end= '')

print()

# utilizando índice para imprimir letras da frase
for c in range(len(frase)):
    print(f'{frase[c]}', end= '')

0 -> 1 -> 2 -> 3 -> 4 -> 5 -> FIM

-=-=-= A frase possui 15 letras -=-=-=
contagem: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 
curso de Python

In [22]:
# letra por letra de trás pra frente

for c in range(len(frase) -1, -1, -1): # o índice começa em 0 e vai até 14, por isso tem de subtrair 1 do tamanho (15)
    print(f'{frase[c]}', end='')

nohtyP ed osruc

In [7]:
# spoiler: forma mais fácil de percorrer uma string

letras = 'curso de Python' # string é uma lista de caracteres

for letra in letras: # usa-se in para percorrer lista
    print(letra, end='')

curso de Python

<mark> Fazer os exercícios do 40 ao 50. </mark>

### Estrutura de Repetição while
Estrutura de repetição com teste lógico

In [2]:
for c in range(1, 10):
    print(c, end=' ')
print('FIM')

# Sabendo o limite, também é possível usar while

c = 1 # contador/início

while c < 10:
    print(c, end=' ')
    c += 1 # passo
print('FIM')

1 2 3 4 5 6 7 8 9 FIM
1 2 3 4 5 6 7 8 9 FIM


In [1]:
# Quando não souber o limite, while é mais eficaz
# Por exemplo, o loop deve finalizar quando o usuário digitar 0:

num = 1

while num != 0: # é necessário um flag para interromper o laço
    num = int(input('Digite um valor: '))
print('FIM')

Digite um valor: 1
Digite um valor: 2
Digite um valor: 4
Digite um valor: 9
Digite um valor: 2
Digite um valor: 0
FIM


In [2]:
# obs: é possível interromper o laço com for utilizando a instrução break

for c in range(1, 5):
    num = int(input('Digite um valor: '))
    if num == 0:
        break # interrompe o laço antes do fim
print('FIM')

Digite um valor: 3
Digite um valor: 4
Digite um valor: 0
FIM


In [17]:
num = 0
resp = 'S'

while resp == 'S': # flag
    num = int(input('Digite um valor: '))
    resp = str(input('Quer continuar? [S/N] ')).upper()
print('FIM')

Digite um valor: 3
Quer continuar? [S/N]s
Digite um valor: 4
Quer continuar? [S/N]s
Digite um valor: 6
Quer continuar? [S/N]s
Digite um valor: 2
Quer continuar? [S/N]n
FIM


In [19]:
num =1
par = impar = 0

while num != 0: # flag
    num = int(input('Digite um valor: '))
    if num != 0: # flag não pode contar como par
        if n % 2 == 0:
            par += 1
        else:
            impar += 1
print(f'Você digitou {par} números pares e {impar} números ímpares')

Digite um valor: 2
Digite um valor: 3
Digite um valor: 5
Digite um valor: 8
Digite um valor: 9
Digite um valor: 13
Digite um valor: 4
Digite um valor: 0
Você digitou 3 números pares e 4 números ímpares


<mark> Fazer os exercícios do 51 ao 59. </mark>

In [13]:
num = soma = 0

while num != 999: # flag
    num = int(input('Digite um número: '))
    soma += num
soma -= 999 # gambiarra para subtrair o flag da soma
print(f'A soma vale {soma}')

num = soma = 0

while num != 999: # flag
    num = int(input('Digite um número: '))
    if num != 999: # modo correto: se não for o flag, soma
        soma += num
print(f'A soma vale {soma}')

Digite um número: 4
Digite um número: 5
Digite um número: 8
Digite um número: 999
A soma vale 17
Digite um número: 4
Digite um número: 5
Digite um número: 8
Digite um número: 999
A soma vale 17


In [17]:
# utilizando a instrução break no laço com while

soma = 0

while True: # será sempre verdade
    num = int(input('Digite um número: '))
    if num == 999: # flag
        break # instrução necessária para sair do loop infinito
    soma += num
print(f'A soma vale {soma}')

Digite um número: 4
Digite um número: 5
Digite um número: 8
Digite um número: 999
A soma vale 17


### Instrução break e continue
Funcionam quando aninhado em laço de repetição **for** ou **while**

In [4]:
print("Entrando no laço...")
for i in range(10):
    print(i, end = ' ')
    if i == 5:
        break # interrompe o laço antes do fim
  # else:
      # pass
print("\nSaindo do laço...")

Entrando no laço...
0 1 2 3 4 5 
Saindo do laço...


In [2]:
print("Entrando no laço...")
i = 0
while(i<10):
    i += 1
    if (i%2 == 0): # número par
        continue # ignora o comando posterior, mas não interrompe o laço, ou seja, salta para o próximo i
    print(i, end =' ')
print("\nSaindo do laço...")

Entrando no laço...
1 3 5 7 9 
Saindo do laço...


<mark> Fazer os exercícios do 60 ao 65. </mark>

## Variáveis compostas

### Estrutura de Tuplas
**As tuplas são imutáveis!**

In [35]:
x1 = ('abc')
x2 = (123)
y1 = ('a', 'b', 'c') # tupla
y2 = (1, 2, 3)
z1 = 'a', 'b', 'c' # a partir do python 3.5 não precisa de parênteses
z2 = 1, 2, 3
print(type(x1))
print(type(x2))
print(type(y1))
print(type(y2))
print(type(z1))
print(type(z2))

<class 'str'>
<class 'int'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>
<class 'tuple'>


In [37]:
y1 = y1.replace('b', 'B') # erro, pois tuplas são imutáveis e este é um método para strings
print(y1)

AttributeError: 'tuple' object has no attribute 'replace'

In [122]:
# As tuplas são semelhantes a vetores, porém no python uma tupla pode conter tipos diferentes

pessoa = ('Icaro', 39, 'M', 83.5)
print(pessoa)
print(type(pessoa))
print(pessoa[0]) # usando índice
print(type(pessoa[0]))
print(pessoa[1]) # usando índice
print(type(pessoa[1]))

('Icaro', 39, 'M', 83.5)
<class 'tuple'>
Icaro
<class 'str'>
39
<class 'int'>


In [1]:
# índice       0           1       2        3
# índice      -4          -3      -2       -1
lanche = ('Hambúrguer', 'Suco', 'Pizza', 'Pudim')
print(lanche)

# fatiamento

print(lanche[1])
print(lanche[-3])
print(lanche[1:3]) # [1,3)
print(lanche[-3:-1]) # [-3,-1)
print(lanche[2:]) # [2,3]
print(lanche[-3:]) # [-3, -1]
print(lanche[:2]) # [0, 2)
print(lanche[:-2]) # [-4, -2)

('Hambúrguer', 'Suco', 'Pizza', 'Pudim')
Suco
Suco
('Suco', 'Pizza')
('Suco', 'Pizza')
('Pizza', 'Pudim')
('Suco', 'Pizza', 'Pudim')
('Hambúrguer', 'Suco')
('Hambúrguer', 'Suco')


In [2]:
print(lanche[-4:]) # [-4:-1]
print(lanche[1::2]) # incremento = 2
print(lanche[1:3:2]) # [1, 3) com incremento de 2
print(lanche[-3::2]) # [-3,-1] com incremento de 2
print(lanche[-1::-2]) # [-1,-4] com decremento de -2, logo -1 e -3 (reverso)

('Hambúrguer', 'Suco', 'Pizza', 'Pudim')
('Suco', 'Pudim')
('Suco',)
('Suco', 'Pudim')
('Pudim', 'Suco')


In [55]:
lanche[1] = 'Refrigerante' # erro, pois tuplas são imutáveis e também não existe um método igual ao de strings 
print(lanche[1])

TypeError: 'tuple' object does not support item assignment

In [3]:
# percorrendo tuplas

for comida in lanche:
    print(f'{comida} ', end='')
print('\nFim')
for comida in lanche[0:3]: # [0, 3)
    print(f'{comida} ', end='')
print('\nFim')
for comida in lanche[::-1]: # reverso
    print(f'{comida} ', end='')
print('\nFim')

Hambúrguer Suco Pizza Pudim 
Fim
Hambúrguer Suco Pizza 
Fim
Pudim Pizza Suco Hambúrguer 
Fim


In [7]:
# outras formas de percorrer tuplas

print(f'A tupla tem {len(lanche)} elementos:')
print('Índices: ', end='')
for cont in range(0, len(lanche)):
    print(cont, end=' ') # mostra índice
print()

print('Valores: ', end='')
for cont in range(0, len(lanche)):
    print(f'{lanche[cont]} ', end='') # mostra valores

A tupla tem 4 elementos:
Índices: 0 1 2 3 
Valores: Hambúrguer Suco Pizza Pudim 

In [95]:
# mostrando índices e valores lado a lado

for cont in range(0, len(lanche)):
    print(f'{cont} - {lanche[cont]}')
print('Fim')

# função enumerate

for pos, comida in enumerate(lanche):
    print(f'{pos} - {comida}')

0 - Hambúrguer
1 - Suco
2 - Pizza
3 - Pudim
Fim
0 - Hambúrguer
1 - Suco
2 - Pizza
3 - Pudim


### Operações com Tuplas

In [8]:
lista = sorted(lanche) # a função cria uma lista ordenada a partir de uma tupla
print(lista)
print(lanche) # a tupla permanece imutável

['Hambúrguer', 'Pizza', 'Pudim', 'Suco']
('Hambúrguer', 'Suco', 'Pizza', 'Pudim')


In [9]:
a = (2, 5, 4)
b = (5, 8, 1, 2)
c = a + b # apenas concatena, e não é igual a: b + a
d = a , b # tuplas dentro de tupla como (a, b)
print(c)
print(d)

(2, 5, 4, 5, 8, 1, 2)
((2, 5, 4), (5, 8, 1, 2))


In [10]:
print(c.count(5)) # método interno no objeto da tupla semelhante a contagem em string, conta os números 5 em c

2


In [12]:
print(c.index(2)) # índice da primeira ocorrência do valor 2

print(c.index(2, 1)) # começa a procurar o 2 partir do índice 1

0
6


In [14]:
#index      0        1         2        3        4        5        6
nomes = ('Icaro', 'Carlos', 'Maria', 'Pedro', 'Maria', 'Joana', 'Maria')
print(nomes.count('Maria'))
print(nomes.index('Maria')) # mostra a primeira ocorrência
print(nomes.index('Maria', 3)) # mostra a primeira ocorrência a partir da posição 3
print(nomes.index('Maria', 5)) # deslocamento

3
2
4
6


In [123]:
del(nomes[0]) #erro, pois tuplas são imutáveis

TypeError: 'tuple' object doesn't support item deletion

In [127]:
compras = ('lápis', 'caneta', 'caderno', 'calculadora')
del (compras) # apaga a tupla
print(compras)

NameError: name 'compras' is not defined

In [8]:
cidades = ('São Paulo', 'Belo Horisonte', 'Brasília')
'Brasília' in cidades # funciona assim como em strings e listas

True

<mark> Fazer os exercícios do 66 ao 71. </mark>

In [57]:
# Tuple comprehension

# tuple(<expression> for <var> in <iterable> {if <condition})

numeros = tuple(range(1, 21))
print(numeros)

pares = tuple(numero for numero in numeros if numero % 2 == 0) # tem de chamar a função tuple, não bastam os parênteses
print(type(pares))
print(pares)

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
<class 'tuple'>
(2, 4, 6, 8, 10, 12, 14, 16, 18, 20)


In [1]:
# combinando tuplas

superior = ('regata', 'camiseta', 'casaco')
inferior = ('short', 'bermuda', 'calça')

vestimenta = tuple((sup, inf) for sup in superior for inf in inferior)

print(vestimenta)

(('regata', 'short'), ('regata', 'bermuda'), ('regata', 'calça'), ('camiseta', 'short'), ('camiseta', 'bermuda'), ('camiseta', 'calça'), ('casaco', 'short'), ('casaco', 'bermuda'), ('casaco', 'calça'))


In [58]:
# Generator comprehension

# (<expression> for <var> in <iterable> if <condition>)

numeros = tuple(range(1, 11))
print(numeros)

pares = (numero for numero in numeros if numero % 2 == 0) # somente entre parênteses
print(type(pares)) # spoiler: é um objeto do tipo generator que possui os métodos mágicos __iter__ e __next__

print(next(pares)) # 2
print(next(pares)) # 4
print(next(pares)) # 6
print(next(pares)) # 8
print(next(pares)) # 10
# print(next(pares)) # erro

(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
<class 'generator'>
2
4
6
8
10


In [74]:
pares = (numero for numero in numeros if numero%2 == 0)

for num in pares: # usando objeto gerador de sequência
    print(num) # gerando um resultado para cada interação
    
# spoiler: no notebook de funções haverá mais estudos sobre funções geradoras

2
4
6
8
10


In [77]:
# usando a função sum()

impares = (numero for numero in numeros if numero%2 != 0)

'''
soma = 0

for num in impares:
    soma += num
print(soma)
'''

print(sum(impares)) # não está na memória, ele gera e soma

25


<mark> Fazer os exercícios do 72 ao 77. </mark>

### Estrutura de Listas
Semelhante às tuplas, porém **as listas são mutáveis**

In [17]:
num = [2, 5, 9, 1]
print(num)

num[2] = 3 # altera valor pelo index
print(num)

del num[1] # apaga valor pelo index
print(num)

# obs.: num[3] = 7 # não adiciona (erro)

[2, 5, 9, 1]
[2, 5, 3, 1]
[2, 3, 1]


In [2]:
# métodos para adicionar valores a uma lista

num = [2, 5, 9, 1]
print(num)

num.append(7) # adiciona no final
print(num)

num.insert(2, 0) # na posição 2 insere o valor 0
print(num)

[2, 5, 9, 1, 7]
[2, 5, 0, 9, 1, 7]


In [1]:
# concatenando listas

lista1 = [2, 5, 9, 1]
lista2 = [10, 13, 20]

lista1 = lista1 + lista2
print(lista1)

# estendendo a lista 

lista1 = [2, 5, 9, 1]
lista2 = [10, 13, 20]

lista1.extend(lista2)
print(lista1)

[2, 5, 9, 1, 10, 13, 20]
[2, 5, 9, 1, 10, 13, 20]


In [21]:
# métodos para remover valores de uma lista

num = [3, 2, 5, 9, 2, 1]
print(num)

num.pop() # remove o último
print(num)

num.pop(2) # remove na posição 2
print(num)

num.remove(2) # remove o 1º valor, neste caso o 2, encontrado da esquerda pra direita
print(num)

letras = ['w', 'c', 'h', 'm', 'c', 'a']
print(letras)
letras.remove('c')
print(letras)
letras.pop(1)
print(letras)

[3, 2, 5, 9, 2, 1]
[3, 2, 5, 9, 2]
[3, 2, 9, 2]
[3, 9, 2]
[3, 2]
['w', 'c', 'h', 'm', 'c', 'a']
['w', 'h', 'm', 'c', 'a']
['w', 'm', 'c', 'a']


In [19]:
# Algoritmo Bubble sort, usado para ordenar lista

num = [9, 2, 5, 1]
n = len(num) # len() funciona igual em strings (que é uma lista de caracteres) e em tuplas

for i in range(n): # [0,4)
    for j in range(n - i - 1): # o -1 é porque o maior vai para o fim e não precisa compará-lo novamente
        if num[j] > num[j+1]:
            num[j], num[j+1] = num[j+1], num[j]
            
'''Explicação: é ineficiente para grandes conjuntos de dados, pois compara 2 a 2
i = 0, j = [0,3) -> num[0] > num[1], logo inverte 9 e 2 = [2, 9, 5, 1]; 
                    num[1] > num[2], logo inverte 9 e 5 = [2, 5, 9, 1];
                    num[2] > num[3], logo inverte 9 e 1 = [2, 5, 1, 9]
i = 1, j = [0,2) -> num[1] > num[2], logo inverte 5 e 1 = [2, 1, 5, 9]
i = 2, j = [0,1) -> num[0] > num[1], logo inverte 2 e 1 = [1, 2, 5, 9]
'''
print(num)

# spoiler: existem algoritmos mais eficientes como Quick sort, Merge sort e Heap sort (notebook de funções)

[1, 2, 5, 9]


In [1]:
# métodos para ordenar lista

num = [2, 5, 9, 1]

num.sort() # altera a lista colocando em ordem crescente
print(num)

num.sort(reverse=True) # altera a lista colocando em ordem decrescente
print(num)

[1, 2, 5, 9]
[9, 5, 2, 1]
[1, 2, 5, 9]


In [9]:
# método para reverter lista a colocando de trás pra frente

num = [2, 5, 9, 1]
print(num)

num.reverse() # se em ordem crescente, passa para ordem decrescente
print(num)

[2, 5, 9, 1]
[1, 9, 5, 2]


In [4]:
# a função sorted também funciona mas gera outra lista (gera uma lista ordenada dos elementos)

num = [2, 5, 9, 1]
sorted(num)

print(num)

num2 = sorted(num)
print(num2)

num3 = sorted(num, reverse=True)
print(num3)

[2, 5, 9, 1]
[1, 2, 5, 9]
[9, 5, 2, 1]


In [10]:
num = [2, 5, 9, 1]
print(num)

num = num[::-1] # recebendo o reverso da própria variável sem uso de método interno da lista
print(num)

[2, 5, 9, 1]
[1, 9, 5, 2]


In [25]:
# criando lista vazia:
lista1 = []
lista2 = list()
print(type(lista1))
print(type(lista2))

# armazenando valores:
lista1.append(5)
lista1.append(9)
lista1.append(4)
print(lista1)

# lendo listas:
for valor in lista1:
    print(f'{valor}..', end='')
print()

for index, valor in enumerate(lista1):
    print(f'Na posição {index}, encontrei o valor {valor}')

<class 'list'>
<class 'list'>
[5, 9, 4]
5..9..4..
Na posição 0, encontrei o valor 5
Na posição 1, encontrei o valor 9
Na posição 2, encontrei o valor 4


In [77]:
# armazenando em uma lista os valores digitados por um usuário

valores = []

for cont in range(0, 5):
    valores.append(int(input('Digite um valor: ')))
print(valores)

Digite um valor: 4
Digite um valor: 8
Digite um valor: 9
Digite um valor: 1
Digite um valor: 3
[4, 8, 9, 1, 3]


In [7]:
# igualando listas

a = [2, 3, 4, 7]
b = a # não faz cópia, faz ligação !!!!

print(f'Lista A: {a}')
print(f'Lista B: {b}')
print()

b[2] = 8

print(f'Lista A: {a}') # veja que também se alterou
print(f'Lista B: {b}')

Lista A: [2, 3, 4, 7]
Lista B: [2, 3, 4, 7]

Lista A: [2, 3, 8, 7]
Lista B: [2, 3, 8, 7]


In [6]:
# fazendo cópia:

a = [2, 3, 4, 7]
b = a[:] # b recebe todos os elementos de a. Não tem ligação.

print(f'Lista A: {a}')
print(f'Lista B: {b}')
print()

b[2] = 8

print(f'Lista A: {a}') # não se alterou
print(f'Lista B: {b}')

Lista A: [2, 3, 4, 7]
Lista B: [2, 3, 4, 7]

Lista A: [2, 3, 4, 7]
Lista B: [2, 3, 8, 7]


In [27]:
# função all()

valores = [True, False, True] 
print(all(valores)) # só retorna True se todos os resultados forem True

numeros = [1,2,3,4,5]
print(all(numeros)) # retorna True porque avalia que todos são diferentes de 0

vazio = []
print(all(vazio)) # retorna True
print()

lista = [2, 4, 6, 8, 10]

todos_pares = all(x % 2 == 0 for x in lista) 
print(todos_pares)

False
True
True

True


In [22]:
primos = []

for valor in range(2,101):
    if all(valor % n != 0 for n in range(2, int(valor ** 0.5) + 1)): # if True
        primos.append(valor)

print(primos)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


In [1]:
# List comprehension

# [<expression> for <var> in <iterable> {if <condition}]

primos = [valor for valor in range(2, 101) if all(valor % n != 0 for n in range(2, int(valor ** 0.5) + 1))] # entre colchetes

print(primos)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


In [3]:
valores_x = [2, 5, 8, 13]
valores_y = [6, 9, 12, 16]

pares_ordenados = [(x,y) for i, x in enumerate(valores_x) for j, y in enumerate(valores_y) if i == j]
# pares_ordenados = list(zip(valores_x, valores_y)) 

print(type(pares_ordenados))
print(pares_ordenados)

<class 'list'>
[(2, 6), (5, 9), (8, 12), (13, 16)]


<mark> Fazer os exercícios do 78 ao 84. </mark>

### Listas Compostas

In [6]:
pessoa = []
pessoa.append('Icaro')
pessoa.append(38)
pessoas = []
pessoas.append(pessoa) # faz ligação entre as estruturas, então se mudar a primeira, muda a segunda
print(pessoa)
print(pessoas)

# erro comum:
pessoa[0] = 'Maria'
pessoa[1] = 22
pessoas.append(pessoa) # aparecerá duplicado, pois alterou o primeiro e adcionou a lista novamente
print(pessoas)

['Icaro', 38]
[['Icaro', 38]]
[['Maria', 22], ['Maria', 22]]


In [7]:
# sem ligação:

dado = []
dado.append('Icaro')
dado.append(38)
pessoas = []
pessoas.append(dado[:]) # faz uma cópia
print(dado)
print(pessoas)

dado[0] = 'Maria'
dado[1] = 22
pessoas.append(dado[:]) # faz uma cópia da estrutura modificada
print(dado)
print(pessoas)

['Icaro', 38]
[['Icaro', 38]]
['Maria', 22]
[['Icaro', 38], ['Maria', 22]]


In [7]:
# acessando valores dentro das listas

matriz_pessoas = [['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]

print(matriz_pessoas)
print(matriz_pessoas[0])
print(matriz_pessoas[0][0]) # lê de fora pra dentro
print(matriz_pessoas[2][1])

[['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]
['João', 19]
João
13


In [8]:
matriz_pessoas = [['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]

print('[nome, idade] = ', end='')
for pessoa in matriz_pessoas:
    print(f'{pessoa} ', end='')
print()

print('Nomes: ', end='')
for pessoa in matriz_pessoas:
    print(f'{pessoa[0]} ', end='')
print()

print('Idades: ', end='')
for pessoa in matriz_pessoas:
    print(f'{pessoa[1]} ', end='')

[nome, idade] = ['João', 19] ['Ana', 33] ['Joaquim', 13] ['Maria', 45] 
Nomes: João Ana Joaquim Maria 
Idades: 19 33 13 45 

In [9]:
matriz_pessoas = [['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]
for pessoa in matriz_pessoas:
    print(f'{pessoa[0]} tem {pessoa[1]} anos de idade')

João tem 19 anos de idade
Ana tem 33 anos de idade
Joaquim tem 13 anos de idade
Maria tem 45 anos de idade


In [18]:
# apesar de método interno lista.sort() não ordenar matriz, a função sorted() ordena, pois gera lista ordenada dos elementos

matriz_pessoas = [['Bia', 19], ['Ana', 33], ['Zé', 13], ['Maria', 45]]

matriz_ordenada = sorted(matriz_pessoas) # ordena pelo índice 0 de cada elemento, ou seja índice 0 de cada lista interna
print(matriz_ordenada)

[['Ana', 33], ['Bia', 19], ['Maria', 45], ['Zé', 13]]


In [10]:
# utilizando a função tabulate do módulo tabulate

from tabulate import tabulate

matriz_pessoas = [['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]
print(tabulate(matriz_pessoas, headers=['Nome', 'Idade'], tablefmt='plain'))

Nome       Idade
João          19
Ana           33
Joaquim       13
Maria         45


In [5]:
# loop em lista de listas (matrizes) para encontrar o maior número

matriz = [[42, 23, 34], [100, 215, 114], [10.1], [98.7], [12.3]]
maior = 0

for linha in matriz:
    for num in linha:
        if num > maior:
            maior = num
            
print(f'Maior número: {maior}')

Maior número: 215


In [13]:
from tabulate import tabulate

matriz_pessoas = list()
dado = list()
totmaior = totmenor = 0

for c in range(0, 5):
    dado.append(str(input('Nome: ')).strip())
    dado.append(int(input('Idade: ')))
    matriz_pessoas.append(dado[:]) # se não fizer cópia, vai apagar tudo
    dado.clear() # limpa a lista dado com nome e idade

print('-'*30)
print(tabulate(matriz_pessoas, headers=['Nome', 'Idade'], tablefmt='grid'))
print('-'*30)

for pessoa in matriz_pessoas:
    if pessoa[1] >= 21:
        print(f'{pessoa[0]} é maior de idade')
        totmaior += 1
    else:
        print(f'{pessoa[0]} é menor de idade')
        totmenor += 1
print('-='*30)
print(f'Temos {totmaior} maiores e {totmenor} menores de idade')

Nome: Raúl
Idade: 92
Nome: Maria
Idade: 77
Nome: Lucca
Idade: 8
Nome: Ana
Idade: 17
Nome: Renata
Idade: 29
------------------------------
+--------+---------+
| Nome   |   Idade |
+========+=========+
| Raúl   |      92 |
+--------+---------+
| Maria  |      77 |
+--------+---------+
| Lucca  |       8 |
+--------+---------+
| Ana    |      17 |
+--------+---------+
| Renata |      29 |
+--------+---------+
------------------------------
Raúl é maior de idade
Maria é maior de idade
Lucca é menor de idade
Ana é menor de idade
Renata é maior de idade
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Temos 3 maiores e 2 menores de idade


<mark> Fazer os exercícios do 85 ao 92. </mark>

### Extra: Comparando strings

In [77]:
# o Python se baseia na tabela ASCII

print('a' == 'b')
print('b' != 'a')
print('a' > 'X')
print('a' > '1')

False
True
True
True


In [70]:
for i in range(122):
    print(str(i) + '-' + chr(i)) # pode usar "+" para concatenar strings

0-
1-
2-
3-
4-
5-
6-
7-
8-
9-	
10-

11-
12-
13-
14-
15-
16-
17-
18-
19-
20-
21-
22-
23-
24-
25-
26-
27-
28-
29-
30-
31-
32- 
33-!
34-"
35-#
36-$
37-%
38-&
39-'
40-(
41-)
42-*
43-+
44-,
45--
46-.
47-/
48-0
49-1
50-2
51-3
52-4
53-5
54-6
55-7
56-8
57-9
58-:
59-;
60-<
61-=
62->
63-?
64-@
65-A
66-B
67-C
68-D
69-E
70-F
71-G
72-H
73-I
74-J
75-K
76-L
77-M
78-N
79-O
80-P
81-Q
82-R
83-S
84-T
85-U
86-V
87-W
88-X
89-Y
90-Z
91-[
92-\
93-]
94-^
95-_
96-`
97-a
98-b
99-c
100-d
101-e
102-f
103-g
104-h
105-i
106-j
107-k
108-l
109-m
110-n
111-o
112-p
113-q
114-r
115-s
116-t
117-u
118-v
119-w
120-x
121-y


### Conjuntos

Coleção desordenada de elementos únicos.

In [38]:
# conjuntos são definidos usando chaves ou a função set()

frutas = {'banana', 'maçã', 'uva', 'uva'} 

print(frutas) # não armazenou duplicidade e pode mostrar em ordem diferente
print(type(frutas))

{'uva', 'banana', 'maçã'}
<class 'set'>


In [41]:
conjunto_vazio = set()
dicionario_vazio = {}

print(type(conjunto_vazio))
print(type(dicionario_vazio)) # não é conjunto

<class 'set'>
<class 'dict'>


In [12]:
# método para adicionar valores

cursos = {'Exatas', 'Humanas', 'Biológicas'}
print(cursos)

cursos.add('Saúde')
print(cursos)

{'Exatas', 'Biológicas', 'Humanas'}
{'Exatas', 'Biológicas', 'Humanas', 'Saúde'}


In [27]:
# métodos para remover valores

cursos = {'Exatas', 'Humanas', 'Biológicas'}
print(cursos)

cursos.remove('Biológicas') # gera erro se o termo não existir
cursos.discard('Saúde') # não gera erro, mesmo que o termo não exista

print(cursos)

cursos.pop() # remove elemento arbitrário, não o último (não é possível selecionar posição, pois não há ordenação)
print(cursos)

cursos = {'Exatas', 'Humanas', 'Biológicas', 'Saúde'}
cursos.clear() # remove todos os elementos
print(cursos) # conjunto vazio

{'Exatas', 'Biológicas', 'Humanas'}
{'Exatas', 'Humanas'}
{'Humanas'}
set()


In [40]:
# método para copiar conjuntos

frutas1 = {'banana', 'maçã', 'uva'}
frutas2 = frutas1.copy() # cópia independente

print(frutas1)
print(frutas2)

frutas1.discard('uva')
frutas1.add('mamão')
frutas2.discard('maçã')
frutas2.add('laranja')
print(frutas1)
print(frutas2)

{'uva', 'banana', 'maçã'}
{'uva', 'banana', 'maçã'}
{'mamão', 'banana', 'maçã'}
{'uva', 'laranja', 'banana'}


In [4]:
# operações com conjuntos: diferença entre conjuntos

numeros = set(range(1,11))
pares = {num for num in range(1,11) if num % 2 == 0} # set comprehension
impares = numeros - pares
print(impares)

cidades_brasileiras = {'Brasília', 'Belo Horizonte', 'Juiz de Fora', 'Ouro Preto', 'Rio de Janeiro', 'Santos', 'São Paulo'}
cidades_mineiras = {'Belo Horizonte', 'Juiz de Fora', 'Ouro Preto'}
cidades_brasileiras_nao_mineiras = cidades_brasileiras - cidades_mineiras # not
print(cidades_brasileiras_nao_mineiras)

{1, 3, 5, 7, 9}
{'Rio de Janeiro', 'Brasília', 'Santos', 'São Paulo'}


In [11]:
# diferença usando método interno

numeros = set(range(1,11))
pares = {num for num in range(1,11) if num % 2 == 0} # set comprehension
impares = numeros.difference(pares) # método interno: difference()
print(impares)

cidades_brasileiras = {'Brasília', 'Belo Horizonte', 'Juiz de Fora', 'Ouro Preto', 'Rio de Janeiro', 'Santos', 'São Paulo'}
cidades_mineiras = {'Belo Horizonte', 'Juiz de Fora', 'Ouro Preto'}
cidades_brasileiras_nao_mineiras = cidades_brasileiras.difference(cidades_mineiras) # not
print(cidades_brasileiras_nao_mineiras)

{1, 3, 5, 7, 9}
{'Rio de Janeiro', 'Brasília', 'Santos', 'São Paulo'}


In [10]:
# operações com conjuntos: interseção

impares = {1, 3, 5, 7, 9}
maiores_5 = {6, 7, 8, 9, 10}
impares_maiores_5 = impares.intersection(maiores_5) # método interno: intersection()
print(impares_maiores_5)

produtos_mais_vendidos = {'cadeira', 'cama box', 'home theater', 'mesa de centro', 'rack', 'sofá'}
produtos_maior_lucro = {'cama box', 'estante', 'home theater', 'mesa de jantar', 'poltrona'}
mais_vendidos_e_maior_lucro = produtos_mais_vendidos.intersection(produtos_maior_lucro)
print(mais_vendidos_e_maior_lucro)

{9, 7}
{'cama box', 'home theater'}


In [22]:
# conjuntos também aceitam o operador &

impares = {1, 3, 5, 7, 9}
maiores_5 = {6, 7, 8, 9, 10}
impares_maiores_5 = impares & maiores_5 
print(impares_maiores_5)

{9, 7}


In [9]:
# operações com conjuntos: união

menores_7 = {1, 2, 3, 4, 5, 6, 7}
maiores_3 = {4, 5, 6, 7, 8, 9, 10}
numeros = menores_7.union(maiores_3) # método interno: union()
print(numeros)

produtos_mais_vendidos = {'cadeira', 'cama box', 'home theater', 'mesa de centro', 'rack', 'sofá'}
produtos_maior_lucro = {'cama box', 'estante', 'home theater', 'mesa de jantar', 'poltrona'}
mais_vendidos_ou_maior_lucro = produtos_mais_vendidos.union(produtos_maior_lucro)
print(mais_vendidos_ou_maior_lucro)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
{'mesa de centro', 'cama box', 'rack', 'mesa de jantar', 'estante', 'poltrona', 'cadeira', 'sofá', 'home theater'}


In [23]:
# conjuntos também aceitam o operador |

menores_7 = {1, 2, 3, 4, 5, 6, 7}
maiores_3 = {4, 5, 6, 7, 8, 9, 10}
numeros = menores_7 | maiores_3 # pipe significa ou
print(numeros)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}


In [19]:
# operações com conjuntos: diferença simétrica = união - intersecção

menores_7 = {1, 2, 3, 4, 5, 6, 7}
maiores_3 = {4, 5, 6, 7, 8, 9, 10}

numeros = menores_7.union(maiores_3) - menores_7.intersection(maiores_3)
print(numeros)

ou_menores_7_ou_maiores_3 = menores_7.symmetric_difference(maiores_3) # método interno: union()
print(ou_maiores_7_ou_maiores_3)

{1, 2, 3, 8, 9, 10}
{1, 2, 3, 8, 9, 10}


In [31]:
# verificando se A está contido em B (se A é subconjunto de B)

a = {3, 5, 7}
b = {2,3,4,5,6,7,8}

a_sub_b = a.issubset(b)
print(a_sub_b)

# verificando se B contém A (se B superconjunto de A)

b_sup_a = b.issuperset(a)
print(b_sup_a)

# verificando se C e D não têm elementos em comum (disjuntos)

c = {2,3,4}
d = {6,7,8}

c_disj_d = c.isdisjoint(d)
print(c_disj_d)

True
True
True


In [32]:
# conjunto é muito utilizado para retirar duplicidade de listas:

hashtags_seg = ['#tiago', '#joao', '#bbb']
hashtags_ter = ['#sarah', '#bbb', '#fiuk']
hashtags_qua = ['#gil', '#thelma', '#lourdes']
hashtags_qui = ['#rafa', '#fora', '#danilo']
hashtags_sex = ['#juliete', '#arthur', '#bbb']

hashtags_semana = hashtags_seg + hashtags_ter + hashtags_qua + hashtags_qui + hashtags_sex
print(hashtags_semana)
print(len(hashtags_semana)) # embora não utilizada, a função len() também funionaria em um conjunto

hashtags_semana = list(set(hashtags_seg + hashtags_ter + hashtags_qua + hashtags_qui + hashtags_sex)) # converte e volta

print(hashtags_semana)
print(len(hashtags_semana))

['#tiago', '#joao', '#bbb', '#sarah', '#bbb', '#fiuk', '#gil', '#thelma', '#lourdes', '#rafa', '#fora', '#danilo', '#juliete', '#arthur', '#bbb']
15
['#joao', '#sarah', '#thelma', '#rafa', '#bbb', '#gil', '#juliete', '#fiuk', '#tiago', '#danilo', '#arthur', '#fora', '#lourdes']
13


<mark> Fazer os exercícios 93 e 94. </mark>

### Estrutura de Dicionários

In [42]:
# dicionários são definidos usando chaves ou a função dict()

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38} # diferente dos conjuntos, o dicionário possui index
print(pessoas)

{'nome': 'Icaro', 'sexo': 'M', 'idade': 38}


In [43]:
pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}
print(pessoas[0]) # erro, pois diferente das listas, o index que seria 0 agora é index literal, neste caso 'nome'

KeyError: 0

In [44]:
# itens, chaves e valores

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

print(f'O {pessoa["nome"]} tem {pessoa["idade"]} anos')

print(pessoa.keys()) # chaves em uma lista
print(pessoa.values()) # valores em uma lista
print(pessoa.items()) # tuplas de chave e valor em uma lista

O Icaro tem 38 anos
dict_keys(['nome', 'sexo', 'idade'])
dict_values(['Icaro', 'M', 38])
dict_items([('nome', 'Icaro'), ('sexo', 'M'), ('idade', 38)])


In [45]:
pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

for k in pessoa.keys(): # percorrendo chaves com método interno
    print(k, end=' ')

print()

for v in pessoa.values(): # percorrendo valores
    print(v, end=' ')

print()

for k, v in pessoa.items(): # chave e valor (não é possível usar a função enumerate)
    print(f'{k}: {v}')

nome sexo idade 
Icaro M 38 
nome: Icaro
sexo: M
idade: 38


In [46]:
pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

lista1 = list(pessoa) # = list(pessoas.keys())
lista2 = list(pessoa.values())
lista3 = list(pessoa.items())

print(lista1)
print(lista2)
print(lista3) # gera uma lista de tuplas

['nome', 'sexo', 'idade']
['Icaro', 'M', 38]
[('nome', 'Icaro'), ('sexo', 'M'), ('idade', 38)]


In [47]:
# alterando volares no dicionário

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

pessoa['nome'] = 'Gustavo'
pessoa['idade'] = 40

for k, v in pessoa.items():
    print(f'{k}: {v}')

print()

# adicinando nova chave e valor no dicionário

pessoa['peso'] = 98.5 

for k, v in pessoa.items():
    print(f'{k}: {v}')

nome: Gustavo
sexo: M
idade: 40

nome: Gustavo
sexo: M
idade: 40
peso: 98.5


In [48]:
# acessando dicionário com o método get()

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

print(pessoa.get('nome', 'valor não encontrado.'))

print(pessoa.get('naturalidade','valor não encontrado.')) # imprime mensagem ao invés do valor None

Icaro
valor não encontrado.


In [52]:
# alterando dicionário com uso do método update()

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

pessoa.update({'nome':'Gustavo'}) 
print(pessoas)

{'idade': 38}


In [53]:
# adicionando valores com o uso do método setdefault()

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

pessoa.setdefault('nome', 'Gustavo') # o valor não é adcionado, pois já existe
pessoa.setdefault('peso', 86.5)

print(pessoas)

{'idade': 38}


In [56]:
# removendo itens do dicionário

pessoa = {'nome':'Icaro', 'sexo':'M', 'idade':38}

del pessoa['sexo']
print(pessoa)

pessoa.pop('nome', 'valor não encontrado') # método pop()

print(pessoa)

{'nome': 'Icaro', 'idade': 38}
{'idade': 38}


In [15]:
# dicionários dentro de lista

brasil = []
estado1 = {'UF':'Rio de Janeiro', 'Sigla':'RJ'}
estado2 = {'UF':'São Paulo', 'Sigla':'SP'}

brasil.append(estado1)
brasil.append(estado2)

print(brasil)
print(brasil[0]['UF'])
print(brasil[1]['Sigla'])

[{'UF': 'Rio de Janeiro', 'Sigla': 'RJ'}, {'UF': 'São Paulo', 'Sigla': 'SP'}]
Rio de Janeiro
SP


In [16]:
# usando a função tabulate do módulo tabulate

from tabulate import tabulate

brasil = []
estado1 = {'UF':'Rio de Janeiro', 'Sigla':'RJ'}
estado2 = {'UF':'São Paulo', 'Sigla':'SP'}

brasil.append(estado1)
brasil.append(estado2)

print(tabulate(brasil, headers='keys', tablefmt='pipe'))

| UF             | Sigla   |
|:---------------|:--------|
| Rio de Janeiro | RJ      |
| São Paulo      | SP      |


In [28]:
# ordenando dicionário com a função sorted()

# sorted(iterável, key=None, reverse=False)

# O parâmetro key (opcional) por padrão (key=None) ordena listas, tuplas e strings com base nos próprios elementos

pessoas = [
    {'nome': 'Alice', 'idade': 25},
    {'nome': 'Bob', 'idade': 20},
    {'nome': 'Carol', 'idade': 30}
]

# não há como comparar dicionário com dicionário para saber qual é maior ou menor, portanto ordena-se por outro critério

pessoas_ordenadas = sorted(pessoas, key= lambda dicio: dicio['idade']) # lambda é uma função anônima
print(pessoas_ordenadas)

# spoiler: lambda será melhor explicada no notebook de funções

[{'nome': 'Bob', 'idade': 20}, {'nome': 'Alice', 'idade': 25}, {'nome': 'Carol', 'idade': 30}]


In [36]:
# percorrendo lista de dicionários

pessoas = [{'Nome': 'Ana', 'Idade': 18}, 
           {'Nome': 'Bia', 'Idade': 50}, 
           {'Nome': 'Carlos', 'Idade': 26}, 
           {'Nome': 'Henrique', 'Idade': 32}
]

for dicio in pessoas[:2]: # apenas os dois primeiros
    print(f'Seu nome é {dicio["Nome"]} e sua idade é {dicio["Idade"]}')

print()

for dicio in pessoas[-2:]: # apenas os dois últimos
    print(f'Seu nome é {dicio["Nome"]} e sua idade é {dicio["Idade"]}')

Seu nome é Ana e sua idade é 18
Seu nome é Bia e sua idade é 50

Seu nome é Carlos e sua idade é 26
Seu nome é Henrique e sua idade é 32


In [19]:
# ordenando matriz por outro índice

from tabulate import tabulate

matriz_pessoas = [['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]

matriz_pessoas_ordenadas = sorted(matriz_pessoas, key= lambda pessoa: pessoa[1], reverse=True) # lambda é uma função anônima

print(tabulate(matriz_pessoas_ordenadas, headers=['Nome', 'Idade'], tablefmt='plain'))

Nome       Idade
Maria         45
Ana           33
João          19
Joaquim       13


In [40]:
# a função dict() também itera sobre matriz

matriz_pessoas = [['João', 19], ['Ana', 33], ['Joaquim', 13], ['Maria', 45]]
dicio = dict(matriz_pessoas) # neste caso, as listas internas só podem possuir 2 elementos

print(dicio)

{'João': 19, 'Ana': 33, 'Joaquim': 13, 'Maria': 45}


In [2]:
# dict comprehension

numeros = [1, 2, 3, 4, 5]

dicionario_quadrados = {num: num ** 2 for num in numeros} # entre chaves
print(dicionario_quadrados)

pessoas = {'Lucca': 8, 'Maria': 77, 'Icaro': 38, 'Alice': 16}

pessoas_maiores_de_idade = {nome: idade for nome, idade in pessoas.items() if idade >= 18}
print(pessoas_maiores_de_idade)

alunos = {'Gustavo':43, 'Ana':54, 'Lucca':96, 'Icaro':73, 'Bia':68, 'Maria':85}

alunos_aprovados = {nome:('Aprovado' if nota >= 70 else 'Reprovado') for nome, nota in alunos.items()}
print(alunos_aprovados)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}
{'Maria': 77, 'Icaro': 38}
{'Gustavo': 'Reprovado', 'Ana': 'Reprovado', 'Lucca': 'Aprovado', 'Icaro': 'Aprovado', 'Bia': 'Reprovado', 'Maria': 'Aprovado'}


<mark> Fazer os exercícios do 95 ao 101. </mark>

## Trabalhando com Arquivos de Texto

Gravando arquivo pelo Jupyter Notebook (a célula abaixo não é Python):

In [34]:
%%writefile noticia.txt
Pessoas ainda não têm ‘devida maturidade’ para entender ChatGPT, diz professor da Uerj
Para especialistas, ferramenta não tem capacidade de tirar empregos de advogados, jornalistas, entre outros profissionais

Overwriting noticia.txt


### Função open()

In [35]:
# abre o arquivo e retorna um objeto (do tipo IO) correspondente. Caso não puder ser aberto, retornará OSError

# default
arquivo = open(file='noticia.txt',
               mode='r', 
               buffering=-1,
               encoding=None,
               errors=None,
               newline=None,
               closefd=True,
               opener=None)

'''
Modos: r = read (default), w = write (cria novo ou sobreescreve, se já existir),
a = append (adiciona conteúdo ao final do arquivo, mas se não existir, o arquivo é criado),
x = exclusive creation (falha se o arquivo já existir)
t = text (será tratado como texto e a codificação padrão será aplicada),
b = binário (leitura de arquivos binários, como imagens, vídeos, etc.)
É possível combinar modos, por exemplo: rb  

buffering: O valor padrão é -1, que significa usar um valor padrão do sistema.
Você pode fornecer outros valores para controlar o comportamento do buffering, como 0 para desabilitar o buffering,
1 para buffering em linha, ou um número inteiro maior para controlar o tamanho do buffer em bytes.

encoding: Este parâmetro especifica a codificação a ser usada ao ler ou gravar o arquivo.
Por padrão, é usado o valor None, o que significa que a codificação padrão do sistema será aplicada.
Exemplos comuns de codificações são 'utf-8', 'latin-1', 'ascii', entre outras.

errors:  Este parâmetro especifica como os erros de codificação devem ser tratados ao ler ou gravar o arquivo.
O valor padrão é None, o que significa que o comportamento padrão do sistema será usado. 
Alguns exemplos de valores são 'strict' (gerar um erro se ocorrer um erro de codificação),
'ignore' (ignorar caracteres inválidos), 'replace' (substituir caracteres inválidos por um caractere de substituição)
e 'backslashreplace' (substituir caracteres inválidos por sequências de escape).

newline: controla a conversão de quebras de linha ao ler ou gravar o arquivo.
O valor padrão é None, o que significa que a conversão de quebras de linha é determinada automaticamente pelo modo de acesso.
Alguns exemplos de valores são '' (não fazer conversão), '\n' (converter para nova linha)
e '\r\n' (converter para retorno de carro seguido de nova linha).

closefd: Este parâmetro é um valor booleano que indica se o arquivo deve ser fechado quando o objeto IO é fechado.
O valor padrão é True, o que significa que o arquivo será fechado. Se você definir esse parâmetro como False,
o arquivo não será fechado quando o objeto IO for fechado.

opener: Este é um objeto personalizado que pode ser usado para abrir o arquivo.
Ele deve ser uma função ou um objeto com um método __call__() que recebe o nome do arquivo e o modo de acesso como argumentos
e retorna um objeto file. Esse parâmetro é menos comum e geralmente não é necessário na maioria dos casos.
'''

print(type(arquivo)) # spoiler: objeto será melhor explicado no notebook de programação orientada a objetos

conteudo = arquivo.read()
print(conteudo) # \n quebra linha, mas não aparece escrito

<class '_io.TextIOWrapper'>
Pessoas ainda não têm ‘devida maturidade’ para entender ChatGPT, diz professor da Uerj
Para especialistas, ferramenta não tem capacidade de tirar empregos de advogados, jornalistas, entre outros profissionais



In [3]:
# É importante lembrar de fechar o arquivo após a leitura/escrita para liberar os recursos do sistema.

arquivo = open(file='noticia.txt', mode='r', encoding='utf8')

'''
Mode: O parâmetro '+' pode ser combinado com outros modos de acesso para habilitar a leitura e escrita simultâneas.

'r+'= Abre o arquivo para leitura e escrita. O arquivo deve existir previamente.

'w+'= Abre o arquivo para leitura e escrita. Se o arquivo já existir, o seu conteúdo será apagado.
Se o arquivo não existir, um novo arquivo vazio será criado.

'a+'= Abre o arquivo para leitura e escrita, adicionando o conteúdo ao final do arquivo sem apagar o conteúdo existente.
Se o arquivo não existir, um novo arquivo vazio será criado.

Esses modos permitem que você leia e escreva no arquivo usando os métodos read(), write(), readline(), writelines(),
entre outros, de forma simultânea.
'''
conteudo = arquivo.readlines()

print(conteudo) # gera uma lista com o \n aparecendo
arquivo.close() # fecha arquivo

['Pessoas ainda não têm ‘devida maturidade’ para entender ChatGPT, diz professor da Uerj\n', 'Para especialistas, ferramenta não tem capacidade de tirar empregos de advogados, jornalistas, entre outros profissionais']


In [4]:
# newline

print(open(file='noticia.txt', encoding='utf8', newline='\r').readlines()) # só separaria se fosse \r

['Pessoas ainda não têm ‘devida maturidade’ para entender ChatGPT, diz professor da Uerj\nPara especialistas, ferramenta não tem capacidade de tirar empregos de advogados, jornalistas, entre outros profissionais']


In [4]:
# buffering padrão: -1

import io

io.DEFAULT_BUFFER_SIZE

8192

In [8]:
# tamanho do arquivo

import os

os.stat('noticia.txt').st_size

215

In [9]:
# encoding padrão usado é o da minha máquina, neste caso: windows 1252 (= cp1252)

import locale

locale.getpreferredencoding()

'cp1252'

In [13]:
# escrevendo em um arquivo utilizando os métodos write() e writelines()

fp = open(file='notícia.txt', mode='w') # escrever os parâmetros explicitamente é uma boa prática
linha = 'Chat GPT: uso de ferramenta de inteligência artificial é analisada por TJMG'
fp.write(linha)
linha = '\n' + 'Objetivo é que sistema auxilie no desenvolvimento de textos das áreas administrativas do tribunal'
fp.write(linha)
fp.close()

arquivo = open(file='notícia.txt', mode='r', encoding='cp1252')
conteudo = arquivo.read() # faz toda iteração e armazena tudo na memória, por isso não é a melhor opção
print(conteudo)
arquivo.close()

fp = open(file='notícia.txt', mode='w') # fp significa file pointer
conteudo = '''Chat GPT pode ajudar na educação?
Diretor administrativo da escola Teia Multicultural explica que a ferramenta pode ajudar a melhorar o aprendizado'''
fp.write(conteudo)
fp.close()

arquivo = open(file='notícia.txt', mode='r')
conteudo = arquivo.read()
print(conteudo)
arquivo.close()

print()

fp = open(file='notícia.txt', mode='w', encoding='utf8')
linhas = ['Construindo soluções baseadas em inteligência artificial generativa' + '\n',
         'Quais são as opções disponíveis e quais os prós e contras de cada uma?']
fp.writelines(linhas) # recebe uma lista
fp.close()

arquivo = open(file='notícia.txt', mode='r', encoding='utf8')
conteudo = arquivo.read()
print(conteudo) # não aparece uma lista, pois utilizaou read() e não readlines()
arquivo.close()

Chat GPT: uso de ferramenta de inteligência artificial é analisada por TJMG
Objetivo é que sistema auxilie no desenvolvimento de textos das áreas administrativas do tribunal

Chat GPT pode ajudar na educação?
Diretor administrativo da escola Teia Multicultural explica que a ferramenta pode ajudar a melhorar o aprendizado

Construindo soluções baseadas em inteligência artificial generativa
Quais são as opções disponíveis e quais os prós e contras de cada uma?



In [6]:
# a função open() produz um objeto iterável

arquivo = open(file='notícia.txt', mode='r', encoding='utf8')
linha1 = arquivo.readline().strip() # readline(), sem s, acrescenta '\n' e o strip() retira espaço e '\n'
linha2 = arquivo.readline().strip()
linha3 = arquivo.readline().strip() # linha vazia
arquivo.seek(0) # retorna pra primeira linha
linha4 = arquivo.readline().strip() # = linha1
arquivo.close()

print(linha1)
print(linha2)
print(linha3)
print(linha4)
print()

# reproduzindo readlines() iterando linha a linha com readline()

arquivo = open(file='notícia.txt', mode='r', encoding='utf8')

conteudo = []
linha = arquivo.readline()
while linha:
    conteudo.append(linha)
    linha = arquivo.readline()

print(conteudo)
arquivo.close()

Construindo soluções baseadas em inteligência artificial generativa
Quais são as opções disponíveis e quais os prós e contras de cada uma?

Construindo soluções baseadas em inteligência artificial generativa

['Construindo soluções baseadas em inteligência artificial generativa\n', 'Quais são as opções disponíveis e quais os prós e contras de cada uma?']


In [5]:
# utilizando o laço for para leitura

arquivo = open(file='notícia.txt', mode='r', encoding='utf8')
conteudo = arquivo.readlines() # lista

for linha in conteudo:
    print(linha.replace('\n','')) # outra forma de remover o \n
    
arquivo.close()

Construindo soluções baseadas em inteligência artificial generativa
Quais são as opções disponíveis e quais os prós e contras de cada uma?


In [6]:
# o cursor para no ponto final da leitura 

arquivo = open(file='notícia.txt', mode='r', encoding='utf8')
print(arquivo.read())
print(arquivo.read()) # não vai imprimir
arquivo.seek(0) # volta o cursor pro início
print(arquivo.read(30)) # lê os primeiros 30 caracteres
print(arquivo.tell()) # mostra a posição do cursor, contando o número de caracteres já percorridos e soma 2 (\n)
print(arquivo.read()) # lê o restante do arquivo a partir do ponto de parada anterior
arquivo.close()

Construindo soluções baseadas em inteligência artificial generativa
Quais são as opções disponíveis e quais os prós e contras de cada uma?


Construindo soluções baseadas 
32
em inteligência artificial generativa
Quais são as opções disponíveis e quais os prós e contras de cada uma?



salarios.csv (Fonte: https://data.cityofchicago.org/)

In [31]:
# dividindo arquivo em linhas

arquivo = open(file='salarios.csv', mode='r', encoding='utf8')
dados = arquivo.read()
arquivo.close()

linhas = dados.split('\n') # lista sem o \n ao contrario do readlines()

print(linhas[0:6]) # cabeçalho + 5 primeiras linhas
print()
print(linhas[-2:]) # dados em branco no final da lista

['Name,Position Title,Department,Employee Annual Salary', '"AARON,  ELVIA J",WATER RATE TAKER,WATER MGMNT,$88967.00', '"AARON,  JEFFERY M",POLICE OFFICER,POLICE,$80778.00', '"AARON,  KARINA",POLICE OFFICER,POLICE,$80778.00', '"AARON,  KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,$84780.00', '"ABAD JR,  VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,$104736.00']

[',,,', '']


In [44]:
# dividindo linhas em colunas

arquivo = open(file='salarios.csv', mode='r', encoding='utf8')
dados = arquivo.read()
arquivo.close()
linhas = dados.split('\n')
linhas.pop() # remove a última linha em branco
linhas.pop() # remove os dados em branco que estavam na penúltima linha

dados_completos = []
for linha in linhas:
    linha_separada = linha.split(',')
    if linha != linhas[0]:
        linha_separada[0] = linha_separada[0] + linha_separada[1] # last name + first name
        linha_separada.pop(1)
    dados_completos.append(linha_separada) # lista composta

print(dados_completos[0:6]) # cabeçalho + 5 primeiras listas

[['Name', 'Position Title', 'Department', 'Employee Annual Salary'], ['"AARON  ELVIA J"', 'WATER RATE TAKER', 'WATER MGMNT', '$88967.00'], ['"AARON  JEFFERY M"', 'POLICE OFFICER', 'POLICE', '$80778.00'], ['"AARON  KARINA"', 'POLICE OFFICER', 'POLICE', '$80778.00'], ['"AARON  KIMBERLEI R"', 'CHIEF CONTRACT EXPEDITER', 'GENERAL SERVICES', '$84780.00'], ['"ABAD JR  VICENTE M"', 'CIVIL ENGINEER IV', 'WATER MGMNT', '$104736.00']]


In [26]:
# usando a função tabulate do módulo tabulate

from tabulate import tabulate

tabulate(dados_completos[1:6], headers=dados_completos[0], tablefmt='html', showindex=True)

'<table>\n<thead>\n<tr><th style="text-align: right;">  </th><th>Name                </th><th>Position Title          </th><th>Department      </th><th>Employee Annual Salary  </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;"> 0</td><td>&quot;AARON  ELVIA J&quot;    </td><td>WATER RATE TAKER        </td><td>WATER MGMNT     </td><td>$88967.00               </td></tr>\n<tr><td style="text-align: right;"> 1</td><td>&quot;AARON  JEFFERY M&quot;  </td><td>POLICE OFFICER          </td><td>POLICE          </td><td>$80778.00               </td></tr>\n<tr><td style="text-align: right;"> 2</td><td>&quot;AARON  KARINA&quot;     </td><td>POLICE OFFICER          </td><td>POLICE          </td><td>$80778.00               </td></tr>\n<tr><td style="text-align: right;"> 3</td><td>&quot;AARON  KIMBERLEI R&quot;</td><td>CHIEF CONTRACT EXPEDITER</td><td>GENERAL SERVICES</td><td>$84780.00               </td></tr>\n<tr><td style="text-align: right;"> 4</td><td>&quot;ABAD JR  VICENTE M&quot;</td><td>CIVIL ENGINEER IV       </td><td>WATER MGMNT     </td><td>$104736.00              </td></tr>\n</tbody>\n</table>'

### Extra: Módulo OS
- Fornece métodos para manipular caminhos de arquivo de forma independente do sistema operacional.
- Oferece funções para verificar a existência de diretórios, criar novos diretórios, navegar pelos diretórios e outras operações.
- Permite verificar e alterar as permissões de arquivo, como permissões de leitura, gravação e execução.

In [18]:
# permitindo que o código funcione corretamente em diferentes sistemas operacionais

import os

diretorio_atual = os.getcwd() # Retorna o diretório de trabalho atual como uma string.
print(diretorio_atual)

arquivo = open(os.path.join('noticia.txt'), mode='r') # caso não estivesse na mesma pasta, poderia usar o caminho de outra
conteudo = arquivo.readlines()
arquivo.close()

print(conteudo)

C:\Users\User\OneDrive\Documentos\GitHub\Introducao-ao-Python\Python-Fundamentos
['Chat GPT-4: inteligência artificial mente para completar tarefa e gera preocupação\n', 'Relatório da OpenAI revela "temores éticos" com comportamentos de "alto risco" da ferramenta']


In [13]:
arquivos = os.listdir('C:/Users/User/OneDrive/Documentos/GitHub/Introducao-ao-Python/Python-Fundamentos') # inverter as barras
print(arquivos)

'''
os.path.exists(caminho) = Verifica se o caminho especificado existe.
os.path.isfile(caminho) = Verifica se o caminho especificado é um arquivo.
os.path.isdir(caminho) = Verifica se o caminho especificado é um diretório.
os.rename(antigo, novo) = Renomeia um arquivo ou diretório
os.chdir(caminho) = Muda o diretório de trabalho para o caminho especificado.
os.mkdir(caminho) = Cria um novo diretório com o caminho especificado.
os.makedirs(caminho) = Cria um diretório completo com todos os subdiretórios no caminho especificado.
os.remove(caminho) = Remove o arquivo.
os.rmdir(caminho) = Remove o diretório vazio.
os.system(comando) = Executa um comando do sistema no shell.
'''

['.gitattributes', '.ipynb_checkpoints', 'arquivos', 'banco.csv', 'calculadora.py', 'DSA-Python-Cap06-02-TXT-CSV-JSON.ipynb', 'DSA-Python-Cap06-07-Zip.ipynb', 'DSA-Python-Cap06-08-Enumerate.ipynb', 'exercício12.txt', 'fatorial.py', 'lib', 'LICENSE', 'moeda.py', 'moeda2.py', 'noticia.txt', 'Python Fundamentos (Exercícios).ipynb', 'Python Fundamentos.ipynb', 'Python Funções (Exercícios).ipynb', 'Python Funções.ipynb', 'Python Módulos e Pacotes (Exercícios).ipynb', 'Python Módulos e Pacotes.ipynb', 'Python POO.ipynb', 'README.md', 'salarios.csv', 'utilidades', '__pycache__']


'\nos.path.exists(caminho) = Verifica se o caminho especificado existe.\nos.path.isfile(caminho) = Verifica se o caminho especificado é um arquivo.\nos.path.isdir(caminho) = Verifica se o caminho especificado é um diretório.\nos.rename(antigo, novo) = Renomeia um arquivo ou diretório\nos.chdir(caminho) = Muda o diretório de trabalho para o caminho especificado.\nos.mkdir(caminho) = Cria um novo diretório com o caminho especificado.\nos.makedirs(caminho) = Cria um diretório completo com todos os subdiretórios no caminho especificado.\nos.remove(caminho) = Remove o arquivo.\nos.rmdir(caminho) = Remove o diretório vazio.\n'

## Tratamento de erros em Python

### Erros e Exceções
- O try e catch são blocos de código utilizados para lidar com exceções durante a execução do programa. O try define um bloco de código onde você suspeita que uma exceção possa ocorrer, enquanto o catch (também conhecido como except) define o bloco de código que trata a exceção caso ela ocorra.

In [5]:
# erros são sintáticos, exceções são erros semânticos

try:
    a = int(input('Numerador: '))
    b = int(input('Denominador: '))
    r = a/b
except: # falhou
    print('Infelizmente tivemos um problema. =(')
else: # deu certo (opcional)
    print(f'O resultado é {r:.1f}')
finally: # imprime sempre
    print('Volte sempre! Muito obrigado!')
    
# Obs.: Não é uma boa prática utilizar o except de forma genérica

Numerador: 5
Denominador: 0
Infelizmente tivemos um problema. =(
Volte sempre! Muito obrigado!


In [6]:
# exceções comuns

vendas = 1342.50
clientes = 0

try:
    vendas_por_cliente = vendas/clientes # operações numéricas impossíveis
    print(vendas_por_cliente)
except ZeroDivisionError:
    print(f'Número de clientes inválido. Espera-se um valor maior que 0 e obteve-se um valor igual a {clientes}')

anos = [2021, 2022, 2023]

try:
    ano_atual = anos[3] # erro de indexação
    print(ano_atual)
except Exception:
    print(f'A lista dos anos é menor que o valor escolhido. Espera-se um valor entre 0 e {len(anos) - 1}')
    
nome = 'Lucca'
idade = 8

try:
    apresentacao = 'O nome dela é ' + nome + ' e ele tem ' + idade + ' anos' # erro de tipos diferentes
    print(apresentacao)
except TypeError:
    idade = str(idade) # tratando o erro com cast
finally:
    apresentacao = 'O nome dele é ' + nome + ' e ele tem ' + idade + ' anos'
    print(apresentacao)

Número de clientes inválido. Espera-se um valor maior que 0 e obteve-se um valor igual a 0
A lista dos anos é menor que o valor escolhido. Espera-se um valor entre 0 e 2
O nome dele é Lucca e ele tem 8 anos


In [3]:
# Multiplos blocos de except (erros específicos)

try:
    a = int(input('Numerador: '))
    b = int(input('Denominador: '))
    r = a/b
except (ValueError, TypeError):
    print('Infelizmente tivemos um problema com os tipos de dados que você digitou. =(')
except ZeroDivisionError:
    print('Não é possível dividir um número por zero')
except KeyboardInterrupt:
    print('O usuário preferiu não informar os dados') # quando clica em stop
except Exception as exc:
    print(exc)

Numerador: 5
Denominador: 0
Não é possível dividir um número por zero


In [2]:
# Capturando o tipo de erro

try:
    a = int(input('Numerador: '))
    b = int(input('Denominador: '))
    r = a/b
except Exception as exc: #falhou
    print(f'O problema encontrado foi {exc.__class__}') # não é pra mostrar isso pro usuário

Numerador: 5
Denominador: 0
O problema encontrado foi <class 'ZeroDivisionError'>


In [49]:
# linhas em colunas 2: iterando linha a linha, assim não precisa carregar tudo na memória

dados = []

arquivo = open(file='salarios.csv', mode='r', encoding='utf8')
linha = arquivo.readline().strip().split(',') # cabeçalho
dados.append(linha)
linha = arquivo.readline().strip() # lê a primeira linha
while linha:
    linha_separada = linha.split(',')
    try:
        linha_separada[0] = f'{linha_separada[0]}, {linha_separada[1].strip()}' # last name + first name
        linha_separada.pop(1)
    except IndexError:
        pass
    except Exception as exc:
        print(exc)
    else:
        dados.append(linha_separada)
    linha = arquivo.readline().strip()
arquivo.close()

print(dados[0:6])
print(dados[-2:]) # dados.pop()

[['Name', 'Position Title', 'Department', 'Employee Annual Salary'], ['"AARON, ELVIA J"', 'WATER RATE TAKER', 'WATER MGMNT', '$88967.00'], ['"AARON, JEFFERY M"', 'POLICE OFFICER', 'POLICE', '$80778.00'], ['"AARON, KARINA"', 'POLICE OFFICER', 'POLICE', '$80778.00'], ['"AARON, KIMBERLEI R"', 'CHIEF CONTRACT EXPEDITER', 'GENERAL SERVICES', '$84780.00'], ['"ABAD JR, VICENTE M"', 'CIVIL ENGINEER IV', 'WATER MGMNT', '$104736.00']]
[['"ZYSKOWSKI, DARIUSZ"', 'CHIEF DATA BASE ANALYST', 'DoIT', '$110352.00'], [', ', '', '']]


In [50]:
# transformando matriz em data frame (lista de dicionários)

data_frame = []
colunas = dados[0][:]
dados.pop(0)

for linha in dados:
    dicio = {}
    for indice, valor in enumerate(linha):
        dicio[colunas[indice]] = valor
    data_frame.append(dicio)
    
print(data_frame[0:5])
print(data_frame[-2:]) # dados.pop()

[{'Name': '"AARON, ELVIA J"', 'Position Title': 'WATER RATE TAKER', 'Department': 'WATER MGMNT', 'Employee Annual Salary': '$88967.00'}, {'Name': '"AARON, JEFFERY M"', 'Position Title': 'POLICE OFFICER', 'Department': 'POLICE', 'Employee Annual Salary': '$80778.00'}, {'Name': '"AARON, KARINA"', 'Position Title': 'POLICE OFFICER', 'Department': 'POLICE', 'Employee Annual Salary': '$80778.00'}, {'Name': '"AARON, KIMBERLEI R"', 'Position Title': 'CHIEF CONTRACT EXPEDITER', 'Department': 'GENERAL SERVICES', 'Employee Annual Salary': '$84780.00'}, {'Name': '"ABAD JR, VICENTE M"', 'Position Title': 'CIVIL ENGINEER IV', 'Department': 'WATER MGMNT', 'Employee Annual Salary': '$104736.00'}]
[{'Name': '"ZYSKOWSKI, DARIUSZ"', 'Position Title': 'CHIEF DATA BASE ANALYST', 'Department': 'DoIT', 'Employee Annual Salary': '$110352.00'}, {'Name': ', ', 'Position Title': '', 'Department': ''}]


In [51]:
# usando a função tabulate do módulo tabulate

from tabulate import tabulate

tabulate(data_frame[0:5], headers='keys', tablefmt='html', showindex=True)

'<table>\n<thead>\n<tr><th style="text-align: right;">  </th><th>Name                </th><th>Position Title          </th><th>Department      </th><th>Employee Annual Salary  </th></tr>\n</thead>\n<tbody>\n<tr><td style="text-align: right;"> 0</td><td>&quot;AARON, ELVIA J&quot;    </td><td>WATER RATE TAKER        </td><td>WATER MGMNT     </td><td>$88967.00               </td></tr>\n<tr><td style="text-align: right;"> 1</td><td>&quot;AARON, JEFFERY M&quot;  </td><td>POLICE OFFICER          </td><td>POLICE          </td><td>$80778.00               </td></tr>\n<tr><td style="text-align: right;"> 2</td><td>&quot;AARON, KARINA&quot;     </td><td>POLICE OFFICER          </td><td>POLICE          </td><td>$80778.00               </td></tr>\n<tr><td style="text-align: right;"> 3</td><td>&quot;AARON, KIMBERLEI R&quot;</td><td>CHIEF CONTRACT EXPEDITER</td><td>GENERAL SERVICES</td><td>$84780.00               </td></tr>\n<tr><td style="text-align: right;"> 4</td><td>&quot;ABAD JR, VICENTE M&quot;</td><td>CIVIL ENGINEER IV       </td><td>WATER MGMNT     </td><td>$104736.00              </td></tr>\n</tbody>\n</table>'

## Gerenciadores de contexto (PEP-343):

### Função with()
with expression [as variable]:<br />
&ensp;&ensp;&ensp;&ensp;&ensp;with-block

In [1]:
# antes do with(), usava-se try e finally

from builtins import list as b_list

try:
    list = 'bananas' # usou a palara reservada list como variável em um contexto
    print(list)
finally:
    list = b_list 
    print(list)

bananas
<class 'list'>


In [11]:
# usando try e finally para abertura e fechamento de arquivos

try:
    fp = open(file='noticia.txt', mode='w')
    linha = 'Chat GPT-4: inteligência artificial mente para completar tarefa e gera preocupação'
    fp.write(linha)
    linha = '\n' + 'Relatório da OpenAI revela "temores éticos" com comportamentos de "alto risco" da ferramenta'
    fp.write(linha)
except IOError:
    print ("Erro: arquivo não encontrado ou não pode ser salvo.")
except Exception as exc:
    print(exc)
else:
    print ("Conteúdo gravado com sucesso!")
finally:
    fp.close() # mesmo que tivesse algum erro, ele fecha o arquivo

print()    

try:
    arquivo = open(file='noticia.txt', mode='r')
    conteudo = arquivo.read()
    print(conteudo)
except IOError:
    print ("Erro: arquivo não encontrado ou não pode ser lido.")
except Exception as exc:
    print(exc)
else:
    print ("\nConteúdo lido com sucesso!")
finally:
    arquivo.close()

Conteúdo gravado com sucesso!

Chat GPT-4: inteligência artificial mente para completar tarefa e gera preocupação
Relatório da OpenAI revela "temores éticos" com comportamentos de "alto risco" da ferramenta

Conteúdo lido com sucesso!


In [3]:
# Mostrando erro e fechando o arquivo

try:
    arquivo = open(file='noticia', mode='r')
    conteudo = arquivo.read()
except IOError:
    print ("Erro: arquivo não encontrado ou não pode ser lido.")
except Exception as exc:
    print(exc)
else:
    print ("\nConteúdo lido com sucesso!")
finally:
    arquivo.close()

Erro: arquivo não encontrado ou não pode ser lido.


In [2]:
# usando with() para abertura e fechamento de arquivos

with open(file='noticia.txt', mode='w') as fp:
    linha = 'Bard: "ChatGPT do Google" é lançado no Brasil; sabia como funciona a IA e como usar'
    fp.write(linha)
    linha = '\n'+'O robô de inteligência artificial, configurado em mais de 40 idiomas, já pode ser usado gratuitamente no país'
    fp.write(linha)
    # fp.close() é feito sozinho ao usar with() e fecha mesmo que ocorra alguma exceção, assim como o finally
    
with open(file='noticia.txt', mode='r') as arquivo:
    conteudo = arquivo.readlines()
    
for linha in conteudo:
    print(linha.strip())

'''
context manager protocol

__enter__ : ele entra em determinado contexto, no qual as coisas vão ser feitas
__exit__ : ele finaliza o contexto e devolve tudo como estava anteriormente
'''

Bard: "ChatGPT do Google" é lançado no Brasil; sabia como funciona a IA e como usar
O robô de inteligência artificial, configurado em mais de 40 idiomas, já pode ser usado gratuitamente no país


'\ncontext manager protocol\n\n__enter__ : ele entra em determinado contexto, no qual as coisas vão ser feitas\n__exit__ : ele finaliza o contexto e devolve tudo como estava anteriormente\n'

In [3]:
# lendo e reescrevendo um arquivo

with open(file='salarios.csv', mode='r', encoding='utf8') as infile:
    cabecalho = infile.readline().strip()
    with open(file='salarios2.csv', mode='w', encoding='utf8') as outfile:
        outfile.write(cabecalho)
        linha = infile.readline().strip() # primeira linha
        while linha:
            linha_separada = linha.split(',')
            try:
                linha_arrumada = '\n'+f'{linha_separada[0]} {linha_separada[1].strip()},{linha_separada[2]},{linha_separada[3]},{linha_separada[4].replace("$","")}'
                outfile.write(linha_arrumada) # gravando linha
            except IndexError:
                pass
            except Exception as exc:
                print(exc)
            linha = infile.readline().strip() # lendo a próxima linha

<mark> Fazer os exercícios do 102 ao 104. </mark>

## Extra: Expressões Regulares com o módulo re
Expressões regulares são padrões usados para combinar ou encontrar ocorrências de sequências de caracteres em uma string. </br>
Documentação: https://docs.python.org/3.9/library/re.html

In [3]:
import re

texto = "Meu e-mail é exemplo@gmail.com e você pode me contatar em outro_email@yahoo.com."

arrobas = len(re.findall("@", texto))  # conta quantas vezes o caracter arroba aparece no texto
print(f'O caractere "@" apareceu {arrobas} vezes no texto.')

'''
Parâmetros:
pattern = padrão de expressão regular que você deseja procurar na string
string = sequência de caracteres em que você deseja procurar o padrão
flags (opcional) = argumento para controlar o comportamento da expressão regular 
'''

palavra = re.findall(r'você (\w+)', texto) # extrai a palavra após a palavra "você" em um texto (gera uma lista)
print(f'A palavra após "você" é: {palavra[0]}')

emails = re.findall(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b', texto) # extrai endereços de e-mail de uma string
print(emails)

# Obs.: O r antes da string que é usado para indicar que a string é uma string literal raw.

O caractere "@" apareceu 2 vezes no texto.
A palavra após "você" é: pode
['exemplo@gmail.com', 'outro_email@yahoo.com']


In [6]:
# re.finditer é similar a re.findall(), mas retorna um iterador que produz objetos de correspondência.

texto = "O aluno estava incrivelmente perdido, mas encontrou a DSA e rapidamente começou a aprender."

for adverbio in re.finditer(r"\w+mente\b", texto): # extrai advérbios da frase
    print(f'{adverbio.start()}-{adverbio.end()}: {adverbio.group(0)}')

15-28: incrivelmente
60-71: rapidamente


## Extra: Manipulando Arquivo CSV com o módulo csv
Comma-Separated Values

In [29]:
import csv

with open(file='numeros.csv', mode='w') as fp:
    writer = csv.writer(fp) # cria o objeto de gravação
    writer.writerow(('nota1', 'nota2', 'nota3')) # cabeçalho
    writer.writerow((63,87,92)) 
    writer.writerow((61,79,76))
    writer.writerow((72,64,91))
    
with open(file='numeros.csv', mode='r', newline = '\r\n') as arquivo:
    conteudo = csv.reader(arquivo) # cria o objeto iterável de leitura, não carrega tudo na memória
    for linha in conteudo: # precisa ser identado (com arquivo aberto)
        print(linha) # poderia fazer append em outra lista, formando uma lista composta

['nota1', 'nota2', 'nota3']
['63', '87', '92']
['61', '79', '76']
['72', '64', '91']


In [27]:
# iterando e jogando na memória (em uma lista)

with open(file='numeros.csv', mode='r', newline = '\r\n') as arquivo:
    conteudo = csv.reader(arquivo) # cria o objeto de leitura
    dados = list(conteudo) # carrega tudo em uma lista
    
print(dados)

[['nota1', 'nota2', 'nota3'], ['63', '87', '92'], ['61', '79', '76'], ['72', '64', '91']]


## Extra: Manipulando Arquivo JSON com o módulo json
Java Script Object Notation

In [8]:
import json

dict_guido = {'nome': 'Guido van Rossum',
              'linguagem': 'Python',
              'similar': ['c','Modula-3','lisp'],
              'users': 1000000}

with open(file='dados.json', mode='w') as fp: # Criando um arquivo Json
    fp.write(json.dumps(dict_guido)) # Converte o dicionário para str e grava
    
with open(file='dados.json', mode='r') as arquivo:
    conteudo = arquivo.read()
    dados = json.loads(conteudo) # converte str para dicionário
    
print(dados)

{'nome': 'Guido van Rossum', 'linguagem': 'Python', 'similar': ['c', 'Modula-3', 'lisp'], 'users': 1000000}


In [4]:
# extraindo dados da internet com urllib.request

import json
from urllib.request import urlopen # from pacote.modulo import função

try:
    resposta = urlopen("http://vimeo.com/api/v2/video/57733101.json").read().decode('utf8') # retorna uma str que parece dict
    dados = json.loads(resposta)[0] # só a primeira parte de uma lista de strings
except urllib.error.URLError:
    print('O site não está acessível no momento.')
except Exception as exc:
    print(exc)
else:
    print ('Título: ', dados['title'])
    print ('URL: ', dados['url'])
    print ('Duração: ', dados['duration'])
    print ('Número de Visualizações: ', dados['stats_number_of_plays'])

Título:  The Good Man trailer
URL:  https://vimeo.com/57733101
Duração:  143
Número de Visualizações:  6618


In [6]:
# utilizando outro módulo: requests

import json
import requests as req

resposta = req.get('http://vimeo.com/api/v2/video/57733101.json') # retorna um objeto
print(f'status code: {resposta.status_code}')

status code: 200


In [7]:
# resposta.text é o atributo do objeto que retorna uma str que parece dict

dados = json.loads(resposta.text)[0] # converte str para dict

print ('Título: ', dados['title'])
print ('URL: ', dados['url'])
print ('Duração: ', dados['duration'])
print ('Número de Visualizações: ', dados['stats_number_of_plays'])

Título:  The Good Man trailer
URL:  https://vimeo.com/57733101
Duração:  143
Número de Visualizações:  6618


In [10]:
# Copiando o conteúdo de um arquivo para outro.

arquivo_fonte = 'dados.json'
arquivo_destino = 'dados.txt'

with open(file=arquivo_fonte, mode='r') as infile:
    conteudo = infile.read()
    with open(file=arquivo_destino, mode='w') as outfile:
        outfile.write(conteudo)
        
# em uma linha: open(arquivo_destino,'w').write(open(arquivo_fonte,'r').read()) 

with open(file='dados.txt', mode='r') as arquivo:
    conteudo = arquivo.read()
    dados = json.loads(conteudo)
    
print(conteudo)

{"nome": "Guido van Rossum", "linguagem": "Python", "similar": ["c", "Modula-3", "lisp"], "users": 1000000}


## Extra: Importando um Dataset com Pandas

In [1]:
import pandas as pd

pd.__version__

'1.4.4'

In [2]:
df = pd.read_csv(filepath_or_buffer='salarios2.csv') # define a primeira linha como cabeçalho por default
df.head()

Name            Position Title        Department  \
0      AARON ELVIA J          WATER RATE TAKER       WATER MGMNT   
1    AARON JEFFERY M            POLICE OFFICER            POLICE   
2       AARON KARINA            POLICE OFFICER            POLICE   
3  AARON KIMBERLEI R  CHIEF CONTRACT EXPEDITER  GENERAL SERVICES   
4  ABAD JR VICENTE M         CIVIL ENGINEER IV       WATER MGMNT   

   Employee Annual Salary  
0                 88967.0  
1                 80778.0  
2                 80778.0  
3                 84780.0  
4                104736.0

In [5]:
# frequência de variáveis qualitativas

print(df['Position Title'].value_counts())

POLICE OFFICER                            9489
FIREFIGHTER-EMT                           1191
SERGEANT                                  1138
FIREFIGHTER                                970
POLICE OFFICER (ASSIGNED AS DETECTIVE)     808
                                          ... 
FOREMAN OF WINDOW WASHERS                    1
SUPVSR OF INFORMATION SERVICES               1
MANAGER OF TAX POLICY                        1
FOREMAN OF ARCHITECTURAL IRON WORKERS        1
MECHANICAL ENGINEER IV                       1
Name: Position Title, Length: 1094, dtype: int64
